### PR TITLE
Issue 679

### PR DESCRIPTION
**Pull request by sgpearse**
Fix for issue #679

### CODE REVIEW

**Comment by shaomeng** [comment]
I included within the FileOperationChecker to report an error because there are multiple ways to fail a file operation. The caller would not know exactly why the operation failed if he only receives a false return value. 
Moreover, I cannot think of a case where a file operation is about to fail, and the caller does NOT want to inform the user. 
Thoughts? @clyne @sgpearse

**Comment by clyne** [comment]
> On Apr 20, 2018, at 12:26 PM, Samuel Li <notifications@github.com> wrote:
> 
> I included within the FileOperationChecker to report an error because there are multiple ways to fail a file operation. The caller would not know exactly why the operation failed if he only receives a false return value.

You can have the checker method return an error string or status flag as an argument. 

> Moreover, I cannot think of a case where a file operation is about to fail, and the caller does NOT want to inform the user.

Sure. You may want to use utilities to explore multiple options. Is there a readable file of some sort in the current working directory? No? How about in the home directory? 

In general things that return status are more useful if they don’t enforce an error reporting mechanism that may not be wanted in all use cases. That’s the model that we use throughout the codebase. 

jc

> Thoughts? @clyne @sgpearse
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

John Clyne
National Center for Atmospheric Research
303.497.1236 (w), 303.809.1922 (c)
clyne@ucar.edu

**Comment by shaomeng** [comment]
John, that is a good point. I've changed the behavior of `FileOperationChecker` so that it 1) does not pop up a window, and 2) it records the latest error message.

**Comment by sgpearse** [comment]
Should be good now.

**Comment by sgpearse** [comment]
Should be good now.

File: apps/vaporgui/MainForm.cpp [hunk 1]
```diff
@@ -1178,6 +1184,11 @@ void MainForm::_fileSaveHelper(string path)
 		return;
 	}
 
+	if (!FileOperationChecker::FileGoodToWrite(fileName)) {
```
Comment (shaomeng) [inline apps/vaporgui/MainForm.cpp]: This "FileGoodToWrite()" check happens AFTER `_controlExec` tries to write the file (L1181). It should, however, happen BEFORE `_controlExec` attempts to write.

File: apps/vaporgui/SettingsEventRouter.cpp [hunk 2]
```diff
@@ -197,43 +198,110 @@ void SettingsEventRouter::_changesPerSaveChanged() {
...
 		QString::fromStdString( sParams->GetAutoSaveSessionFile() ),
-        tr("Vapor 3 Session Files (*.vs3)")
-    );
+		tr("Vapor 3 Session Files (*.vs3)")
+	);
 
-	if (! fileName.isEmpty()) 
-    {
+	bool goodToWrite = FileOperationChecker::FileGoodToWrite(fileName); 
+	if (! fileName.isEmpty() && goodToWrite) {
```
Comment (shaomeng) [inline apps/vaporgui/SettingsEventRouter.cpp]: Testing both if `fileName` is empty and if it's good to write is redundant. That is, if a fileName were empty, it won't pass the "GoodToWrite" test.

File: lib/render/Visualizer.cpp [hunk 3]
```diff
@@ -998,11 +998,13 @@ int Visualizer:: captureImage(string filename)
     else // write png files
     {
 		FILE* test = fopen((const char*)filename.c_str(), "wb");
-        fclose( test );
-		if (!test) {
+		if (!test)
```
Comment (shaomeng) [inline lib/render/Visualizer.cpp]: `FileOperationChecker` should be used here to simplify and modulate code.

Comment (sgpearse) [reply lib/render/Visualizer.cpp]: FileOperationChecker is in the GUI, and the Visualzier is in the render library.  Can't use it yet!

File: apps/vaporgui/SettingsEventRouter.cpp [hunk 4]
```diff
@@ -197,43 +198,109 @@ void SettingsEventRouter::_changesPerSaveChanged() {
...
+	);
 
-	if (! fileName.isEmpty()) 
-    {
+	bool goodToWrite = FileOperationChecker::FileGoodToWrite(fileName); 
+	if (goodToWrite) {
 		sParams->SetAutoSaveSessionFile(fileName.toStdString());
 		_saveSettings();
 	}
+	else (!goodToWrite) {
```
Comment (shaomeng) [inline apps/vaporgui/SettingsEventRouter.cpp]: I don't think this line will pass the compiler...

File: lib/render/Visualizer.cpp (hunk 3 above)
Comment (shaomeng) [reply lib/render/Visualizer.cpp]: that's true...

**Pull request merged**